feat(android): add custom headers to all HTTP request paths

yaker-gh opened the pull request:
## Brief summary

Completes the custom HTTP headers implementation for the Android app. The app already had UI scaffolding (`CustomHeadersModal.vue`) and a `customHeaders` field on `ServerConnectionConfig`, but headers were never actually sent with requests. This PR wires them through every HTTP path so the app works behind reverse proxies that require authentication headers (Cloudflare Access, Authelia, Authentik, etc.).

## Which issue is fixed?

Related to discussion: https://github.com/advplyr/audiobookshelf-app/discussions/1313
Closes: https://github.com/advplyr/audiobookshelf-app/issues/254

## Pull Request Type

- **Platform:** Android only
- **Scope:** Frontend (Vue/JS) and Backend (Kotlin)

## In-depth Description

### Commit 1: Core plumbing (`feat(android): support custom HTTP headers for reverse proxy auth`)

**Android (Kotlin):**

| File | Change | Why |
|------|--------|-----|
| `ApiHandler.kt` | Inject custom headers into `getRequest()`, `postRequest()`, `patchRequest()` | All API calls funnel through these methods — single injection point covers all endpoints |
| `DeviceManager.kt` | Add `customHeaders` getter property | Exposes headers from current `ServerConnectionConfig` to all Kotlin code |
| `AbsDatabase.kt` | Sync `customHeaders` when updating *existing* server configs | Previously only set on new configs — editing headers on an existing server had no effect |
| `DownloadItemPart.kt` | Add custom headers to Android `DownloadManager` requests | External-storage downloads use the system DownloadManager, not OkHttp |
| `InternalDownloadManager.kt` | Accept + apply custom headers to OkHttp download requests | Internal-storage downloads go through a separate OkHttp client |
| `DownloadItemManager.kt` | Pass `DeviceManager.customHeaders` to `InternalDownloadManager` | Wiring between header source and download pipeline |
| `PlaybackSession.kt` | try-catch around local cover bitmap decoding | Prevents crash when cover image decode fails (encountered during testing) |

**Frontend (Vue/JS):**

| File | Change | Why |
|------|--------|-----|
| `ServerConnectForm.vue` | Add "Custom Headers" button to UI; pass headers to ping, status check, and socket connect | Users need to configure headers AND they need to be used during initial connection |
| `plugins/nativeHttp.js` | Add custom headers to all `CapacitorHttp` requests | JS-layer native HTTP calls were missing headers entirely |

### Commit 2: Fix remaining HTTP paths (`fix: add CF Access custom headers to all HTTP paths`)

During real-world testing behind Cloudflare Access, several HTTP paths were found to be missing custom headers, causing:
- **Constant logouts** — token refresh requests were blocked by CF Access, so the app thought the token was invalid
- **Broken ebook/PDF rendering** — pdfjs fetches content through its own HTTP layer without custom headers
- **Broken real-time sync** — Socket.IO WebSocket connections were blocked by CF Access

| File | Change | Why |
|------|--------|-----|
| `plugins/axios.js` | Request interceptor injects custom headers into all axios requests | Axios is used alongside nativeHttp for many API calls — was completely missing headers |
| `plugins/nativeHttp.js` | Add headers to `refreshAccessToken()`; pass `serverConnectionConfig` through call chain | Token refresh hitting `/auth/refresh` without CF headers → blocked → forced logout |
| `store/user.js` | Add headers to Vuex `refreshToken` action | Duplicate token refresh code path (Vuex store action) — same logout bug |
| `components/readers/PdfReader.vue` | Add custom headers to pdfjs `httpHeaders` config | PDF/ebook viewer has its own HTTP layer that doesn't go through axios or nativeHttp |
| `plugins/server.js` | Accept `customHeaders` in `connect()`, pass as Socket.IO `extraHeaders` | WebSocket was connecting without auth headers — blocked by proxy |
| `layouts/default.vue` | Pass `customHeaders` when reconnecting socket on app resume | Auto-reconnect used old `connect()` signature without headers |

### Commit 3: UX fix (`fix: show edit form immediately without re-validating server connection`)

| File | Change | Why |
|------|--------|-----|
| `ServerConnectForm.vue` | `editServerConfig()` now shows the form directly instead of calling `submit(true)` first | Previously, editing a saved server config required the server to be reachable. If the server was down or the address changed, you couldn't edit the config. Now matches `newServerConfigClick()` behavior. |

### Commit 4: Ebook reader fix (`fix: enable CapacitorHttp to route ebook requests through native HTTP`)

The ebook readers (epub, comic, mobi, PDF) use axios/XHR for binary data fetches (`responseType: 'arraybuffer'`/`'blob'`). With `CapacitorHttp.enabled: false`, these XHR requests execute in the WebView and are subject to CORS enforcement. Behind reverse proxies like Cloudflare Access, the CORS preflight `OPTIONS` request is blocked because preflight requests cannot carry custom authentication headers per the CORS spec.

| File | Change | Why |
|------|--------|-----|
| `capacitor.config.json` | Set `CapacitorHttp.enabled: true` | Patches global `XMLHttpRequest`/`fetch` to route through native HTTP, bypassing CORS while preserving binary response types |
| `android/.../capacitor.config.json` | Same change in Android assets copy | Android build reads from this copy |

**Why this is safe now:** This was previously disabled because Capacitor 4's native HTTP broke binary responses (returning base64 strings instead of ArrayBuffer/Blob). Capacitor 5.7+ (PR [#6818](https://github.com/ionic-team/capacitor/pull/6818)) fixed this for GET requests by using a proxy URL approach — the native layer handles the HTTP request but the WebView's real XHR processes the response, preserving proper binary types. The app is on Capacitor 7.

### Known limitations

- **iOS:** No custom header support at all — `ServerConnectionConfig.swift` doesn't have a `customHeaders` property. Out of scope for this PR.
- **ExoPlayer HLS streaming:** Only sends `Authorization` header, not custom headers. Direct play uses token-in-URL so it works regardless.
- **Cover images via `<img>` tags:** Browser limitation — `<img>` elements can't carry custom headers. Works if server doesn't require auth for cover endpoints.

## How have you tested this?

- Configured Cloudflare Access with Service Token authentication (`CF-Access-Client-Id` + `CF-Access-Client-Secret`) in front of audiobookshelf server
- Added headers via the Custom Headers modal in the app
- Verified: server ping, login, library browsing, audiobook streaming, downloads (both internal and external storage), ebook/PDF reading, real-time sync via Socket.IO
- Verified token refresh works without triggering logout
- Verified headers persist across app restarts
- Verified headers update correctly when editing an existing server connection
- Verified ebook readers (epub, comic, mobi, PDF) load correctly behind CF Access after enabling CapacitorHttp
- Verified no regression in audio playback, navigation, or socket connections after enabling CapacitorHttp

## Screenshots
<img width="864" height="1939" alt="Screenshot_20251229-231947" src="https://github.com/user-attachments/assets/bd1325e0-90c5-47b5-8b23-6d965992adad" />
<img width="864" height="1939" alt="Screenshot_20251229-231904" src="https://github.com/user-attachments/assets/1f8dba01-8486-453d-b8fc-7546f046e9cc" />
<img width="864" height="1939" alt="Screenshot_20251229-231856" src="https://github.com/user-attachments/assets/331b04ca-7df9-4a06-a1ed-1516c5cd8802" />
<img width="864" height="1939" alt="Screenshot_20251229-231816" src="https://github.com/user-attachments/assets/830ca520-0b5c-4e2c-be39-4b6abcf50763" />

MichaelKopt commented:
Please merge it. For those who use Audiobookshelf with Cloudflare, it is very useful.

Gathaeryx commented:
@advplyr any chance this could be merged in the next beta release? There only seems to be one release every two months, and the one from two weeks ago didn't include it. This the most discussed issue at the moment (after the iOS app), so it seems useful for plenty of people. If any additional testing has to be done to get this merged, I am willing to help.

nichwall commented:
Independent testing of PRs is helpful to indicate that the PR is ready to review. The current priority is on the web client rewrite and there are over 100 open PRs between the server and apps, so it may still be a bit.

If this APK works for you, you can also sideload the APK in the meantime.

MichaelKopt commented:
> If this APK works for you, you can also sideload the APK in the meantime

The issue is getting the APK. I don't have time to set up a build on my local machine, but if I have a beta APK with this feature, I can help with testing.

3gfvgFWESAs44545 commented:
Happy to test as well if an apk can be produced by someone?

yaker-gh commented:
I can produce the APK. I've been running my fork ever since I added this change. I'll need a couple days then I'll post the link here.

yaker-gh commented:
There you go! https://gitlab.com/yaker/audiobookshelf-app/-/releases

Tested on a Pixel 9 running GrapheneOS and it works as expected.

nichwall commented:
I think you may have pushed something incorrectly, there are now over 2000 commits for this PR

yaker-gh commented:
> I think you may have pushed something incorrectly, there are now over 2000 commits for this PR

Yikes. Yes, I think so too.

nichwall commented:
If you can revert back to where it was we can approve workflows for the PR soon (TM). The workflows will automatically build APK for PRs already, it just doesn't happen automatically for new contributors.

yaker-gh commented:
Ooookay. Back to something reasonable.

yaker-gh commented:
I missed _a lot_ of stuff when first implementing this. It works, but it has a few quirks that I need to fix. I'll be updating this PR with quite a few changes in the coming days.

Commit:
fix: add CF Access custom headers to all HTTP paths

Token refresh, axios interceptor, PDF reader, and Socket.IO
were missing custom headers from server connection config.
This caused constant logouts (refresh blocked by CF Access)
and broken ebook/PDF rendering behind Cloudflare Access.

Commit:
feat(android): support custom HTTP headers for reverse proxy auth

Adds support for configuring custom HTTP headers per server connection,
enabling audiobookshelf to work behind reverse proxies that require
authentication headers (e.g., Cloudflare Access, Authelia, Authentik).

Changes:
- Add customHeaders field to server connection config and DeviceManager
- Inject custom headers into all Android native HTTP requests (API, downloads)
- Inject custom headers into JS-layer nativeHttp requests
- Add custom headers UI to server connection form
- Persist custom headers in device database

Commit:
fix: show edit form immediately without re-validating server connection

Previously editServerConfig() called submit(true) and only showed the form
if the server was reachable. This prevented editing configs for unreachable
servers. Now matches newServerConfigClick() behavior: show form, hide auth,
clear errors.

Commit:
fix: enable CapacitorHttp to route ebook requests through native HTTP

Ebook readers (epub, comic, mobi, pdf) use axios/XHR which triggers CORS
preflight requests. Behind reverse proxies like Cloudflare Access, these
preflights are blocked because they can't carry custom auth headers.

Enabling CapacitorHttp patches the global XHR/fetch to route through
native HTTP, bypassing CORS while preserving binary response types
(arraybuffer/blob) for ebook libraries. This was broken in Capacitor 4
but fixed in 5.7+ via proxy URL approach for GET requests (PR #6818).

yaker-gh commented:
@nichwall, I think that's all of it!

yaker-gh committed:
Merge branch 'advplyr:master' into add-custom-headers

3gfvgFWESAs44545 commented:
Do we need testers for this yet? Can I test this feature?
